iina/NowPlayingInfoManager.swift

```diff
@@ -0,0 +1,620 @@
+        center.nowPlayingInfo = nil
+        center.playbackState = .stopped
+        isActive = false
+        log("Ended Now Playing session")
+      }
+      return
+    }
+    
+    // Because showing artwork is a complex operation there is an internal setting that can be
+    // changed to disable this feature should any problems with it be discovered.
+    if Preference.bool(for: .enableNowPlayingArtwork) {
+      // If the media item has changed then if artwork is being displayed it needs to be abandoned.
+      // NOTE this MUST be done before copying the nowPlayingInfo array below as abandonArtwork
+      // works directly on nowPlayingInfo.
+      if activePlayer.info.isNetworkResource {
+        abandonArtwork()
+      }
+      if let url = activePlayer.info.currentURL {
+        if url != artworkURL {
+          if artworkURL != nil {
+            log("Abandoning artwork due to media item changing", level: .verbose)
+          }
+          abandonArtwork()
+          artworkURL = url
+        }
+      } else {
+        abandonArtwork()
+      }
+      // If the best kind of artwork (front cover) is not being displayed then try and update
+      // to a better kind of artwork.
+      if artworkKind.need(.frontCover), !activePlayer.info.isNetworkResource,
+         let url = activePlayer.info.currentURL {
+        updateArtwork(activePlayer, url, activePlayer.info.videoTracks)
+      }
+    } else {
+      log("Showing cover artwork is disabled", level: .verbose)
+    }
+    
+    var info = center.nowPlayingInfo ?? [String: Any]()
+    if withTitle {
+      if activePlayer.currentMediaIsAudio == .isAudio {
+        info[MPNowPlayingInfoPropertyMediaType] = MPNowPlayingInfoMediaType.audio.rawValue
+        let (title, album, artist) = activePlayer.getMusicMetadata()
+        info[MPMediaItemPropertyTitle] = title
+        info[MPMediaItemPropertyAlbumTitle] = album
+        info[MPMediaItemPropertyArtist] = artist
+      } else {
+        info[MPNowPlayingInfoPropertyMediaType] = MPNowPlayingInfoMediaType.video.rawValue
+        info[MPMediaItemPropertyTitle] = activePlayer.getMediaTitle(withExtension: false)
+        info.removeValue(forKey: MPMediaItemPropertyAlbumTitle)
+        info.removeValue(forKey: MPMediaItemPropertyArtist)
+      }
+    }
+
+    let duration = PlayerCore.lastActive.info.videoDuration?.second ?? 0
+    let time = activePlayer.info.videoPosition?.second ?? 0
+    
+    // When playback is paused Now Playing expects the playback rate to be set to zero. If this is
+    // not done the slider in Now Playing may incorrectly display a playback time of zero.
+    let paused = activePlayer.info.state == .paused
+    let speed = paused ? 0 : activePlayer.info.playSpeed
+
+    info[MPMediaItemPropertyPlaybackDuration] = duration
+    info[MPNowPlayingInfoPropertyElapsedPlaybackTime] = time
+    info[MPNowPlayingInfoPropertyPlaybackRate] = speed
+    info[MPNowPlayingInfoPropertyDefaultPlaybackRate] = 1
+
```

Comment:
What about adding `MPNowPlayingInfoPropertyAssetURL` here as well? Didn't notice any difference but the system might use this value.
```suggestion
    info[MPNowPlayingInfoPropertyAssetURL] = activePlayer.info.currentURL

```

Reply:
Also didn't notice any difference after adding `MPNowPlayingInfoPropertyChapterCount` and `MPNowPlayingInfoPropertyChapterNumber`, but those're also something we can add.

Reply:
I'll add the setting of `MPNowPlayingInfoPropertyAssetURL` and then correct it to not be set if `currentURL` is `nil`.

There is a _lot_ that still needs to be done with respects to this code. This PR is focused on the artwork for the Now Playing module. Some of the missing settings like `MPNowPlayingInfoPropertyChapterCount` I don't think matter to the Now Playing module. _However_ I am expecting they do matter if you try and use Siri to control IINA. Originally I was going to open that can of worms for 1.4.0. I'm guessing there are only a few things that need fixing. But currently I was thinking of postponing looking into this as I still have several issues I need to address for beta2. Should I look into this?

Reply:
Let's focus on b2 for now.